### ISSUE TITLE
[Sapling] decide how to derive μ uniquely

### ISSUE THREAD

**Issue by daira** [zcash/zcash#2847]
We need a unique μ for each Sapling Output in order to prevent Faerie Gold attacks.

In Sprout, the corresponding value hSig was derived from randomSeed, nf<sup>old</sup><sub>1..N<sup>old</sup></sub>, and joinSplitPubKey. Of these only the nullifiers were guaranteed to be unique. In Sapling, Outputs can be present without any Spends, and so we don't necessarily have any nullifiers to use for this purpose. There are two obvious approaches:

1. Maintain a separate global set of previously used μ values.
2. Add a dummy nullifier to be used when there is no Spend in the transaction.

**Comment by daira** [comment]
At the Greenwich meeting, we decided that μ should be an EdJubjub public key. Validators maintain a set of used μ values, to ensure that it is unique. The private key corresponding to μ is then used to sign the transaction hash, so that it isn't possible to take a μ from someone else's transaction and block their transaction from being mined.

**Comment by daira** [comment]
Closing since we are no longer going to use a unique μ. The replacement is to derive ρ from both cm and its position in the commitment Merkle tree, as described in https://github.com/zcash/zcash/issues/2277#issuecomment-366528976 .

**Comment by daira** [comment]
> Closing since we are no longer going to use a unique μ. The replacement is to derive ρ from both cm and its position in the commitment Merkle tree, as described in [#2277 (comment)](https://github.com/zcash/zcash/issues/2277#issuecomment-366528976) .

Incidentally, for any cryptoarchaeologists trying to reconstruct the rationale behind the Sapling design, I consider this to have been a mistake. It was very inconvenient for the note commitment to be dependent on the position in the tree. For Orchard we went back to using a ρ that is not position-dependent (which we were able to do easily because the Action design lets ρ be equal to the nullifier in the same Action).
